fhir-server/src/main/java/au/csiro/pathling/QueryHelpers.java

```diff
@@ -141,81 +119,137 @@ public static IdAndValue updateGroupingColumns(@Nonnull final ParserContext cont
     check(right.getIdColumn().isPresent());
-    final Column joinCondition = leftId.equalTo(right.getIdColumn().get());
+    Column joinCondition = leftId.equalTo(right.getIdColumn().get());
+    if (additionalCondition.isPresent()) {
+      joinCondition = joinCondition.and(additionalCondition.get());
+    }
     final Dataset<Row> target = selectJoinTarget(right.getDataset(), left);
     return left.join(target, joinCondition, joinType.getSparkName());
   }
 
   /**
-   * @param left A {@link Dataset}
-   * @param leftId The ID {@link Column} in the left Dataset
-   * @param right Another Dataset
-   * @param rightId The ID column in the right Dataset
-   * @param joinType A {@link JoinType}
-   * @return A new {@link Dataset}
+   * Joins a {@link Dataset} to another Dataset, using resource identity columns.
+   * <p>
+   * This should not be used in contexts where there may be grouping columns.
+   *
+   * @param left a {@link Dataset}
+   * @param leftId the ID {@link Column} in the left Dataset
+   * @param right another Dataset
+   * @param rightId the ID column in the right Dataset
+   * @param joinType a {@link JoinType}
+   * @return a new {@link Dataset}
    */
   @Nonnull
   public static Dataset<Row> joinOnId(@Nonnull final Dataset<Row> left,
       @Nonnull final Column leftId, @Nonnull final Dataset<Row> right,
       @Nonnull final Column rightId, @Nonnull final JoinType joinType) {
-    // Don't do unnecessary joins between identical datasets.
-    if (left.equals(right)) {
-      return left;
-    }
     final Column joinCondition = leftId.equalTo(rightId);
     final Dataset<Row> target = selectJoinTarget(right, left);
     return left.join(target, joinCondition, joinType.getSparkName());
   }
 
   /**
-   * Returns a list of the first N columns within a {@link Dataset}.
+   * Joins two {@link FhirPath} expressions matching the reference (on the left) to the resource
+   * identity (on the right).
    *
-   * @param dataset the Dataset to get the columns from
-   * @param numberOfColumns the number of columns to take
-   * @return a list of {@link Column} objects
+   * @param left a {@link FhirPath}
+   * @param right a {@link FhirPath}
+   * @param joinType a {@link JoinType}
+   * @return a new {@link Dataset}
    */
   @Nonnull
-  public static List<Column> firstNColumns(@Nonnull final Dataset<Row> dataset,
-      final int numberOfColumns) {
-    return Arrays.asList(dataset.columns()).subList(0, numberOfColumns)
-        .stream()
-        .map(dataset::col)
-        .collect(Collectors.toList());
+  public static Dataset<Row> joinOnReference(@Nonnull final FhirPath left,
+      @Nonnull final FhirPath right, @Nonnull final JoinType joinType) {
+    @Nullable final Column reference = left.getValueColumn().getField("reference");
+    checkNotNull(reference);
+    final Column rightId = checkPresent(right.getIdColumn());
+    final Column joinCondition = reference.equalTo(rightId);
```

Comment:
Maybe call the the version of `joinOnRefence` once `rightId` is obtained.

Reply:
See c97c0a716ac60c2410c9c924f5ba2e9afbc977bc.